File: IOSPlayer/KitTest/KitTest/JoyStick.m
//
//  JoyStick.m
//  KitTest
//
//  Created by XYZHENU on 2020/9/6.
//  Copyright © 2020 jizan. All rights reserved.
//

#import "JoyStick.h"

@interface JoyStick()

@property (nonatomic, strong) UIImageView *stickBgView;
@property (nonatomic, strong) UIImageView *stickView;
@property (nonatomic, strong) void(^callback)(CGFloat,CGFloat);
@end

@implementation JoyStick

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        self.userInteractionEnabled = YES;
        [self initView:frame];
    }
    return self;
}

- (void)initView:(CGRect)frame
{
    self.stickBgView = [[UIImageView alloc]initWithFrame:CGRectMake(0, 0, frame.size.width, frame.size.height)];
    [self addSubview:self.stickBgView];
    [self.stickBgView.layer setMasksToBounds:YES];
    [self.stickBgView.layer setCornerRadius:frame.size.width/2];
    
    self.stickView = [[UIImageView alloc] initWithFrame:CGRectMake(frame.size.width/4, frame.size.height/4, frame.size.width/2, frame.size.height/2)];
    [self addSubview:self.stickView];
    
    self.stickView.backgroundColor = [UIColor redColor];
    self.stickBgView.backgroundColor = [UIColor greenColor];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    self.stickBgView.frame = self.bounds;
}

- (void)setCallBack:(void(^)(CGFloat,CGFloat))callback {
    self.callback = callback;
}


- (CGPoint)callinelength:(CGPoint)point {
    CGFloat centrex = self.frame.size.width/2;          //圆心X
    CGFloat centrey = self.frame.size.height/2;         //圆心Y
    CGFloat radius = self.frame.size.width/2;           //半径
    CGFloat x;              //坐标系X
    CGFloat y;              //坐标系Y
    
    x = point.x - centrex;
    y = centrey - point.y;
    
    float current_radius =  sqrtf(x*x + y*y);           //计算改点到圆心的距离
    if(current_radius > radius)
    {
        float circlex = fabs(x) / current_radius * radius;
        float circley = fabs(y) / current_radius * radius;
        if(x < 0 && y > 0)
        {
            x = centrex - circlex;
            y = centrey - circley;
        }
        else if(x > 0 && y > 0)
        {
            x = centrex + circlex;
            y = centrey - circley;
        }
        else if(x < 0 && y < 0)
        {
            x = centrex - circlex;
            y = centrey + circley;
        }
        else if (x > 0 && y < 0)
        {
            x = centrex + circlex;
            y = centrey + circley;
        }
        return CGPointMake(x, y);
    }
    else
    {
        return CGPointMake(point.x, point.y);
    }
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    UITouch *touch = [touches anyObject];
    CGPoint pointBegan = [touch locationInView:self];
    [self updatePosition:[self callinelength:pointBegan]];
}

- (void)touchesMoved:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    UITouch *touch = [touches anyObject];
    CGPoint pointBegan = [touch locationInView:self];
    [self updatePosition:[self callinelength:pointBegan]];
}

- (void)touchesEnded:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event {
    [self updatePosition:CGPointMake(self.frame.size.width/2, self.frame.size.height/2)];
}

- (void)updatePosition:(CGPoint)position {
    self.stickView.center = position;
    if (self.callback) {
        CGFloat radius = self.frame.size.width/2;
        CGFloat xRate = (position.x - radius)/radius;
        CGFloat yRate = (position.y - radius)/radius;
        self.callback(xRate, yRate);
    }
}

@end

File: IOSPlayer/XYRemoter/XYRemoterKit/UI/JoyStick.swift
//
//  File.swift
//  XYRemoterKit
//
//  Created by XYZHENU on 2020/9/13.
//  Copyright © 2020 XYZHENU. All rights reserved.
//

import UIKit

@objc protocol JoyStickReceiver {
    func stickMove(stick:JoyStickView, x:Float, y:Float)
}

class JoyStickView: UIView {
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupUI()
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        setupUI()
    }
    var strokeColor = UIColor.white
    
    private let stickPoint = UIImageView(frame: CGRect(x: 0, y: 0, width: 44, height: 44))
    private func setupUI() {
        isUserInteractionEnabled = true
        backgroundColor = UIColor.clear
        layer.masksToBounds = false
        stickPoint.backgroundColor = strokeColor
        stickPoint.alpha = 0.7
        stickPoint.layer.cornerRadius = stickPoint.frame.size.width / 2
        addSubview(stickPoint)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        updateStickPosition()
    }
    
    override func draw(_ rect: CGRect) {
        super.draw(rect)

        let context = UIGraphicsGetCurrentContext()
        context?.saveGState()
        
        context?.setShadow(offset: CGSize.zero, blur: 4, color: UIColor.red.cgColor)
        
        let radius = rect.size.width / 2
        let circlePath = UIBezierPath(arcCenter: CGPoint(x: radius, y: radius), radius: radius - 4, startAngle: 0, endAngle: CGFloat(Float.pi * 2), clockwise: true)
        strokeColor.setStroke()
        circlePath.stroke()
        
        context?.restoreGState()
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        updateStick(location: positionCaculate(point: touch.location(in: self)))
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        updateStick(location: positionCaculate(point: touch.location(in: self)))
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else { return }
        updateStick(location: positionCaculate(point: touch.location(in: self)))
        resetStick()
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        resetStick()
    }
    
    private func positionCaculate(point:CGPoint) -> (CGPoint, CGPoint) {
        
        let centrex = self.frame.size.width / 2;          //圆心X
        let centrey = self.frame.size.height / 2;         //圆心Y
        let radius = self.frame.size.width / 2;           //半径
        var x:CGFloat = 0;              //坐标系X
        var y:CGFloat = 0;              //坐标系Y
        
        x = point.x - centrex;
        y = point.y - centrey;
        
        let current_radius =  CGFloat(sqrtf(Float(x*x + y*y)));           //计算点到圆心的距离
        if(current_radius > radius)
        {
            let rate = radius / current_radius
            x = x * rate;
            y = y * rate;
            
            return (CGPoint(x: centrex + x, y: centrey + y), CGPoint(x: x / radius, y: -y / radius));
        }
        else
        {
            return (point, CGPoint(x: x / radius, y: -y / radius));
        }
    }
    
    weak var receiver: JoyStickReceiver?
    private func updateStick(location: (position:CGPoint, rate:CGPoint)) {
        stickPoint.center = location.position
        receiver?.stickMove(stick: self, x: Float(location.rate.x), y: Float(location.rate.y))
    }
    private func resetStick() {
        self.receiver?.stickMove(stick: self, x: 0, y: 0)
        UIView.animate(withDuration: 0.3) {
            self.updateStickPosition()
        }
    }
    private func updateStickPosition() {
        let centrex = self.frame.size.width / 2
        let centrey = self.frame.size.height / 2;
        self.stickPoint.center = CGPoint(x: centrex, y: centrey)
    }
    
    func moveStick(x:Float, y:Float) {
        if fabsf(x) < 0.01, fabsf(y) < 0.01 {
            UIView.animate(withDuration: 0.3) {
                self.updateStickPosition()
            }
        } else {
            let radius = self.frame.size.width / 2
            let centrex = radius
            let centrey = self.frame.size.height / 2
            self.stickPoint.center = CGPoint(x: centrex + CGFloat(x) * radius, y: centrey + CGFloat(-y) * radius)
        }
    }
}
